### PR TITLE
Don't remove PASE connection until after commissioning complete

### LINKED ISSUE #16766: BLE should not appear in CASE, causes binary incompatibility
#### Problem
`CASEServer` has a conditional build dependency in APIs on the inclusion of BLE features:

```
CHIP_ERROR ListenForSessionEstablishment(Messaging::ExchangeManager * exchangeManager, TransportMgrBase * transportMgr,
#if CONFIG_NETWORK_LAYER_BLE
                                             Ble::BleLayer * bleLayer,
#endif
                                             SessionManager * sessionManager, FabricTable * fabrics,
                                             Credentials::GroupDataProvider * responderGroupDataProvider);
```

This means that to support the feature where BLE is required, different build must be done and the library may be binary incompatible.

Furthermore, all of this BLE dependency that goes all over, is just for "closing the commissioning link" on CASE establishment, which is not actually required to be done at that layer:

```
CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
{
    ReturnErrorCodeIf(ec == nullptr, CHIP_ERROR_INVALID_ARGUMENT);

    // Mark any PASE sessions used for commissioning as stale.
    // This is a workaround, as we currently don't have a way to identify
    // secure sessions established via PASE protocol.
    // TODO - Identify which PASE base secure channel was used
    //        for commissioning and drop it once commissioning is complete.
    mSessionManager->ExpireAllPairings(kUndefinedNodeId, kUndefinedFabricIndex);

#if CONFIG_NETWORK_LAYER_BLE
    // Close all BLE connections now since a CASE handshake has been initiated.
    if (mBleLayer != nullptr)
    {
        ChipLogProgress(Discovery, "CASE handshake initiated, closing all BLE Connections");
        mBleLayer->CloseAllBleConnections();
    }
#endif
```

 #### Proposed Solution

- We should remove the BLE dependency from the entire CASE API, and make BLE get closed either by:
  - On device: Handling commissioning complete if not already closed (which has its own event handling hooks that have access to BLE layer)
  - On device: Handling end of commissioning window or expiry of fail-safe
  - On commissioner: close BLE session right after CommissioningComplete succeeds, if not already closed, so as to not rely on the peripheral doing it.

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
Commissioner closes the pase session too early - if we're commissioning, we need to hold onto it until the commissioning complete command response is received because the spec indicates that we need to go back to network configuration over pase if that command is unsuccessful. 
Also stop the case server from prematurely closing the pase session. Since the case server is no longer handling the PASE removal, we can also remove the BLE component. 
Also remove BLE component from CommissioneeDeviceProxy since it's completely unused right now.
Fixes #16766

#### Change overview
- set breadcumb on arm failsafe
- hook up the breadcrumb attribute so it can be read
- remove BLE from CASEServer and CommissioneeDeviceProxy
- Don't remove pase connections until after commissioning complete

#### Testing
- extended commissioning_failure tests to simulate a failure after operational connection and before commissining complete.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -556,6 +556,16 @@ CommissioneeDeviceProxy * DeviceCommissioner::FindCommissioneeDevice(const Trans
...
 void DeviceCommissioner::ReleaseCommissioneeDevice(CommissioneeDeviceProxy * device)
 {
+    // TODO: Call CloseSession here see #16440 and #16805 (blocking)
+
+#if CONFIG_NETWORK_LAYER_BLE
+    if (mSystemState->BleLayer() != nullptr && device->GetDeviceTransportType() == Transport::Type::kBle)
+    {
+        // We only support one BLE connection, so if this is BLE, close it
+        ChipLogProgress(Discovery, "Closing all BLE connections");
+        mSystemState->BleLayer()->CloseAllBleConnections();
```
Comment (msandstedt) [inline src/controller/CHIPDeviceController.cpp:566]: This is a path where it would be possible to actually close the specific connection with the connection object.  If we can, shouldn't we?  The fact we can't distinguish BLE connections by address shouldn't block our ability to do this.

**Review by woody-apple — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp:566]: Makes sense - would you mind if I did this in a follow-up? This is a straight move of the logic from the CASEServer into the commissioner. LMK if that works for you and I can open an issue to track.

Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:566]: Yes, thank you.

**Review by msandstedt — APPROVED** [review]

**Comment by cecille** [comment]
Draft while I'm sorting out the test issues with the breadcrumb - it's not getting initialized correctly, and I think the expecatations with the darwin test are incorrect.

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 2]
```diff
@@ -97,6 +97,9 @@ CHIP_ERROR GeneralCommissioningAttrAccess::Read(const ConcreteReadAttributePath
     case SupportsConcurrentConnection::Id: {
         return ReadSupportsConcurrentConnection(aEncoder);
     }
+    case Breadcrumb::Id: {
+        return aEncoder.Encode(DeviceLayer::DeviceControlServer::DeviceControlSvr().GetBreadcrumb());
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:101]: Breadcrumb is supposed to be non-volatile.  Why are we doing something other than just storing it in the attribute store in the normal way?

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 3]
```diff
@@ -252,5 +256,6 @@ bool emberAfGeneralCommissioningClusterSetRegulatoryConfigCallback(app::CommandH
 
 void MatterGeneralCommissioningPluginServerInitCallback()
 {
+    DeviceLayer::DeviceControlServer::DeviceControlSvr().SetBreadcrumb(0);
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:259]: It should already be getting initialized to 0 in the attribute store...

File: src/protocols/secure_channel/CASEServer.cpp [hunk 4]
```diff
@@ -58,22 +52,6 @@ CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
...
-    // secure sessions established via PASE protocol.
-    // TODO - Identify which PASE base secure channel was used
-    //        for commissioning and drop it once commissioning is complete.
-    mSessionManager->ExpireAllPairings(kUndefinedNodeId, kUndefinedFabricIndex);
-
-#if CONFIG_NETWORK_LAYER_BLE
-    // Close all BLE connections now since a CASE handshake has been initiated.
-    if (mBleLayer != nullptr)
-    {
-        ChipLogProgress(Discovery, "CASE handshake initiated, closing all BLE Connections");
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASEServer.cpp:72]: We still need to do this on the device side (but not in this code), right? See "Proposed solution" section of https://github.com/project-chip/connectedhomeip/issues/16766#issue-1184914983

Generally, it seems like we should do the BLE bits in a separate PR from the PASE session bits here....

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 2 above)
Comment (cecille) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:102]: This is just following through with the previous code that used the device control server to store the breadcrumb from the commands. I know there was slack discussion about moving this over to the attribute store, but IMO that should be a separate PR>

File: src/protocols/secure_channel/CASEServer.cpp (hunk 4 above)
Comment (cecille) [reply src/protocols/secure_channel/CASEServer.cpp:72]: Are you requesting a change here? Or just that the change be separated into another PR? The BLE changes are handled in a separate commit.

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:102]: That's fine, yes.  We should just make sure that work is tracked and happens....

File: src/protocols/secure_channel/CASEServer.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.cpp:72]: I would prefer the BLE bits in a separate PR, because right now they are regressing behavior (not actually doing the BLE bits on the server, where we do the right now) and I don't want to hold up the other parts on the BLE part...

Comment (cecille) [reply src/protocols/secure_channel/CASEServer.cpp:72]: Can you clarify the regression? This code block is just moved out to the device commissioner so we don't close the PASE connection as soon as we start initiating CASE. The logic in the CASE server is incorrect and blocks the ability to recover from errors using the PASE connection if we fail to establish CASE.

Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.cpp:72]: In the current setup (which has some problems, as you note), the idea is that on the _commissionee_ once CASE is established the BLE connection will be closed.  This code is not "supposed" to be running on the commissioner at all, because the commissioner wasn't even running a CASEServer when this code was added.  

So moving it to device commissioner totally changes who is closing what, right?

I can see that you need to change the BLE stuff in CASEServer if we're going to not close the BLE connection on either side until after commissioning is complete.  What we should be doing is something like what https://github.com/project-chip/connectedhomeip/issues/16766 describes.

1) The _commissionee_ closes it at the right point (when commissioning is complete, or the fail-safe expires, or the commissioning window closes and there is no existing PASE session).
2) The _commissioner_ closes it at the right point (releasing commissionee device seems ok).  And even then, closing _all_ commiss

I guess given that this CASEServer bit blocks you it might be OK to implement just item 2 for now and then file a followup for item 1, but we need to make sure it happens.

Comment (cecille) [reply src/protocols/secure_channel/CASEServer.cpp:72]: OK, I see what you're saying.

The commissioning window manager already closes out all the BLE connections on a failed attempt and we do expire the pase pairings on commissioning complete, but the BLE connections don't get closed. If we close the BLE connections on commissioning complete, that seems like it would solve the issue. Thoughts?

Comment (bzbarsky-apple) [reply src/protocols/secure_channel/CASEServer.cpp:72]: Yes, that would solve the issue related to CASE.

Looks like we already close BLE connections on session establishment error (as you said) or fail-safe expiration.

The one place that would be left is when the commissioning window timer expires before a PASE session is established.  We don't arm the "initial" fail-safe timer until PASE establishment, so it does not help here.  This should definitely be a separate follow-up for this PR.

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 2 above)
Comment (cecille) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:102]: https://github.com/project-chip/connectedhomeip/issues/17515

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 3 above)
Comment (cecille) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:282]: See above comment - should be addressed in separate PR.

File: src/controller/CHIPDeviceController.cpp [hunk 5]
```diff
@@ -1523,16 +1544,21 @@ void DeviceCommissioner::OnDeviceConnectedFn(void * context, OperationalDevicePr
...
+    else
     {
-        commissioner->ReleaseCommissioneeDevice(commissionee);
+        if (commissioner->mPairingDelegate != nullptr)
+        {
+            commissioner->mPairingDelegate->OnPairingComplete(CHIP_NO_ERROR);
+        }
+        // Only release the pase session if we're not commissioning. If we're commissioning, we're going to hold onto that pase
+        // session until we send the commissioning complete command just in case it fails and we need to go back to the pase
+        // connection to re-setup the network. This is unlikely, given that we just connected over the operational network, but is
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1555]:
```suggestion
        // Only release the PASE session if we're not commissioning. If we're commissioning, we're going to hold onto that PASE
        // session until we send the commissioning complete command just in case it fails and we need to go back to the PASE
        // connection to re-setup the network. This is unlikely, given that we just connected over the operational network, but is
```

File: src/protocols/secure_channel/CASEServer.cpp [hunk 6]
```diff
@@ -60,15 +54,6 @@ CHIP_ERROR CASEServer::InitCASEHandshake(Messaging::ExchangeContext * ec)
 {
     ReturnErrorCodeIf(ec == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
 
-#if CONFIG_NETWORK_LAYER_BLE
-    // Close all BLE connections now since a CASE handshake has been initiated.
-    if (mBleLayer != nullptr)
-    {
-        ChipLogProgress(Discovery, "CASE handshake initiated, closing all BLE Connections");
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASEServer.cpp:67]: Still needs the followup for the "commissioning window timer expires before a PASE session is established" case, right?

**Review by bzbarsky-apple — APPROVED** [review]

File: src/protocols/secure_channel/CASEServer.cpp (hunk 6 above)
Comment (cecille) [reply src/protocols/secure_channel/CASEServer.cpp:67]: Tracking issue here: https://github.com/project-chip/connectedhomeip/issues/17508

File: src/protocols/secure_channel/CASEServer.cpp (hunk 4 above)
Comment (cecille) [reply src/protocols/secure_channel/CASEServer.cpp:67]: added in next commit, tracking issue for no PASE connection is here: https://github.com/project-chip/connectedhomeip/issues/17508
